pkg/revdial/e2e_test.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revdial
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+	"net/url"
+	"sync"
+	"testing"
+	"time"
+)
+
+func setup(t *testing.T) (*http.Client, string, func()) {
+	t.Helper()
+	backend := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello world")
+	}))
+	backend.EnableHTTP2 = true
+	backend.StartTLS()
+
+	// public server
+	pool := NewReversePool()
+	publicServer := httptest.NewUnstartedServer(pool)
+	publicServer.EnableHTTP2 = true
+	publicServer.StartTLS()
+
+	// private server
+	l, err := NewListener(publicServer.Client(), publicServer.URL, "d001")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// reverse proxy queries to an internal host
+	url, err := url.Parse(backend.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	proxy.Transport = backend.Client().Transport
+	server := &http.Server{Handler: proxy}
+	go server.Serve(l)
+
+	// client
+	// wait for the reverse connection to be established
+	time.Sleep(1 * time.Second)
+	stop := func() {
+		l.Close()
+		server.Close()
+		publicServer.Close()
+		backend.Close()
+	}
+	return publicServer.Client(), publicServer.URL + "/proxy/d001/", stop
+
+}
+
+func Test_e2e(t *testing.T) {
```

Comment:
move these into test/e2e? Or is this actually an integration test?

Reply:
integration,